File: 백준/Gold/15686. 치킨 배달/치킨 배달.cc
#include <vector>
#include <iostream>
#include <deque>
#include <set>
#include <algorithm>
using namespace std;

pair <int,int> pi;

int N, M;
int res = 0;

vector<vector<int>> MAP;
vector<pair<int,int>> H;
vector<pair<int,int>> CK;
vector<pair<int,int>> B;

void back(int idx){
    if (B.size() == M){
        int cnt = 0;
        for (int i = 0; i < H.size(); i++)
        {
            int a = 0;
            for (int j = 0; j < B.size(); j++)
            {
                if (a == 0){
                    a = abs(H[i].first-B[j].first) + abs(H[i].second -B[j].second);
                }
                else{
                    a = min(a,abs(H[i].first-B[j].first) + abs(H[i].second -B[j].second));
                }
            }
            cnt += a;
            if (res != 0 && cnt > res){
                return;
            }
        }
        if (res == 0){
            res = cnt;
        }
        else{
            res = min(res,cnt);
        }
    }

    for (int i = idx; i < CK.size(); i++)
    {
        B.emplace_back(CK[i].first,CK[i].second);
        back(i+1);
        B.pop_back();
    }
    
}

int main() {
	ios::sync_with_stdio(0); cin.tie(0);
    cin >> N >> M;
    MAP.resize(N+1,vector<int>(N+1));
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            cin >> MAP[i][j];
            if (MAP[i][j] == 1){
                H.emplace_back(i,j);
            }
            else if (MAP[i][j] == 2){
                CK.emplace_back(i,j);
            }
        }
    }
    back(0);
    cout << res;
}

File: 백준/Gold/15686. 치킨 배달/치킨 배달.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Algorithm
{
    public static void Main(string[] args)
    {
        int[] input = Console.ReadLine().Split().Select(int.Parse).ToArray();
        int N = input[0];
        int M = input[1];
        int res = int.MaxValue;
        List<Tuple<int, int>> Home = new List<Tuple<int, int>>();
        List<Tuple<int, int>> Chicken = new List<Tuple<int, int>>();
        List<int> clist = new List<int>();

        for (int i = 1; i <= N; i++)
        {
            int[] row = Console.ReadLine().Split().Select(int.Parse).ToArray();
            for (int j = 1; j <= N; j++)
            {
                if (row[j - 1] == 1)
                {
                    Home.Add(new Tuple<int, int>(i, j));
                }

                if (row[j - 1] == 2)
                {
                    Chicken.Add(new Tuple<int, int>(i, j));
                }
            }
        }

        void ChooseChicken(int idx)
        {
            if (clist.Count == M)
            {
                CheckDistance();
                return;
            }
            for (int i = idx; i < Chicken.Count; i++)
            {
                clist.Add(i);
                ChooseChicken(i+1);
                clist.RemoveAt(clist.Count - 1);
            }
        }

        void CheckDistance()
        {
            int cnt = 0;
            foreach (Tuple<int, int> h in Home)
            {
                int tmp = int.MaxValue;
                foreach (int i in clist)
                {
                    tmp = Math.Min(tmp, (Math.Abs(h.Item1 - Chicken[i].Item1) + Math.Abs(h.Item2 - Chicken[i].Item2)));
                }
                cnt += tmp;
                if (cnt >= res) break;
            }
            res = Math.Min(res, cnt);
        }

        ChooseChicken(0);
        Console.WriteLine(res);
    }
}